lib/semian/lru_hash.rb

```diff
@@ -0,0 +1,63 @@
+class LRUHash
+  extend Forwardable
+  def_delegators :@table, :size, :count, :empty?, :keys, :clear, :delete
+  attr_reader :table, :minimum_time_in_lru
+
+  class NoopMutex
+    def synchronize(*)
+      yield
+    end
+  end
+
+  def initialize(minimum_time_in_lru:)
+    @minimum_time_in_lru = minimum_time_in_lru
+    @table = {}
+    @lock =
+      if Semian.thread_safe?
+        Mutex.new
+      else
+        NoopMutex.new
+      end
+  end
+
+  def set(key, resource)
+    @lock.synchronize do
+      delete(key)
+      @table[key] = resource
+    end
+    clear_unused_resources
+    resource.updated_at = Time.now
+  end
+
+  def get(key)
+    @lock.synchronize do
+      if found = delete(key)
+        @table[key] = found
+      end
+      @table[key]
+    end
+  end
+
+  def []=(key, resource)
+    set(key, resource)
+  end
+
+  def [](key)
+    get(key)
+  end
+
+  private
+
+  def clear_unused_resources
+    @table.take(2).each do |key, resource|
```

Comment:
After discussing with @jpittis, I've realized that if we have 100 resources and all of them are in use, we will still go through all the resources. Since the `get` puts the value at the back of the hash, it's safe to assume that the un-used values are at the front and will get cleared first. The reason I break on `resource.updated_at + minimum_time_in_lru > Time.now` is if we hit that minimum time, it means all the other resources are also going to be true, therefore, we do not need to keep going through the hash. Let me know if that makes sense, I will run some benchmark with and without `take(2)` to see the difference.